### PR TITLE
feat(internal/database): badger v3 and memory implementations

### PR DESCRIPTION
**Pull request by qdm12**
## Changes

:information_source: This is only the standalone database implementation. I need the in-memory implementation for #2831 which in turn unblocks #2981 which will unblock the full replacement of chaindb. This doesn't affect the existing codebase for now.

- [x] `internal/database` shared (return) interfaces and sentinel errors
- [x] `internal/database/memory`
  - [x] implementation
  - [x] integration test
  - [x] full unit test coverage
  - [x] thread safety test
- [x] `internal/database/badger` - based on badger/v3
  - [x] implementation
  - [x] integration test
  - [x] full unit test coverage
  - [x] thread safety test
- [x] Add race tests to CI

## Tests

```sh
go test github.com/ChainSafe/gossamer/internal/database/...
```

## Issues

Related #1973 / will create an issue for replacing the database implementation

## Primary Reviewer

@timwu20

### CODE REVIEW

File: internal/database/badger/writebatch.go
Comment (qdm12) [inline internal/database/badger/writebatch.go:22]: here we use the badger db writer batch instead of writing our own (chaindb one is faulty on top of that)

File: internal/database/memory/database.go
Comment (qdm12) [inline internal/database/memory/database.go:12]: The point of the in-memory database implementation package is to force the internal/database code to be flexible for additional implementations.

On top of that, I would suggest us to use this one instead of the in-memory badger database, so the badger database is only on disk.

File: internal/database/memory/writebatch.go
Comment (qdm12) [inline internal/database/memory/writebatch.go:23]: Note this fixes the current issue in the chaindb batch, where the order of operation is kept.

File: internal/database/badger/database.go
Comment (EclesioMeloJunior) [inline internal/database/badger/database.go:65]: Could be simplified to wrap any error that happens
```suggestion
        if err != nil {
          fmt.Errorf("getting key 0x%x: %w", key, err)
        }
	return value, err
```

File: internal/database/badger/database.go
Comment (EclesioMeloJunior) [inline internal/database/badger/database.go:72]:
```suggestion
		return txn.Set(key, value)
```

File: internal/database/badger/settings.go
Comment (EclesioMeloJunior) [inline internal/database/badger/settings.go:15]: Are there any problems with using `Path: string` with the default value being `""`?

**Review by EclesioMeloJunior — CHANGES_REQUESTED** [review]

File: internal/database/badger/settings.go
Comment (qdm12) [reply internal/database/badger/settings.go:15]: Well `""` is a valid path (current directory) and is also the default.
I use a string pointer so we can know when the field is left unset.
I need to write a blog post sometime to explain the whole reasoning behind this, but this should explain it more or less: https://github.com/qdm12/gosettings#settings-struct

> Each component has a settings struct, where the zero value of a field should be INVALID. For example, if the value 0 is allowed for a field, then it must be an *int field. On the contrary, you could have an int field if the zero value 0 is not valid. The reasoning behind this is that you want the zero Go value to be considered as 'unset field' so that the field value can be defaulted, merged with or overridden by another settings struct.

Comment (qdm12) [reply internal/database/badger/settings.go:15]: ... never mind, `""` is not valid and crashes the badger open function, so changing this to a non-pointer string!

File: internal/database/badger/helpers.go
Comment (qdm12) [inline internal/database/badger/helpers.go:6]: This function fixes a kinda critical problem that exists for tables in chaindb

**Review by kishansagathiya — COMMENTED** [review]
Looks good overall. 

Yet to look at test files.

File: internal/database/memory/database.go
Comment (qdm12) [reply internal/database/memory/database.go:15]: It could if we add a config field for the database implementation later.
Not that useful I have to admit, but I would say it serves as a good example for another database implementation (as in it's a 'dumb' base just using Go std types).

File: internal/database/memory/writebatch.go
Comment (qdm12) [reply internal/database/memory/writebatch.go:36]: > Are you saying that currently chaindb does not keep order of operations?

Yes 😄 

https://github.com/ChainSafe/chaindb/blob/f900218c88f853b64138c628c2761383f4180dd3/database.go#L89-L90

> Wouldn't that mess up things while executing them?

Yes 😄

File: internal/database/badger/settings.go
Comment (qdm12) [reply internal/database/badger/settings.go:17]: Actually the point of this is to detect when it's left unset, since `false` is not 'unset'. That way we can set defaults as we want with a method.

More details on https://github.com/qdm12/gosettings#settings-struct I have been using this for about a year, and it has been quite scalable and nice to use.

File: internal/database/memory/database.go
Comment (qdm12) [reply internal/database/memory/database.go:85]: > Is this getting used anywhere other than tests?

Both implementations are not used anywhere in this PR. The badger implementation will be used in production code and in tests in the next PR. The in memory implementation could be used in tests, but really we should test using the in-memory badger implementation. So at the end of the day, that memory implementation is really just a base to serve for another implementation (zero dependency, just Go code with what should be expected in terms of implementation)

File: internal/database/memory/table.go
Comment (qdm12) [reply internal/database/memory/table.go:6]: They are similar for all methods but `NewWriteBatch()`. The problem is the writeBatch implementation for both packages is quite different, so I don't think this can be abstracted away really. Unless we would use our own in-memory write batch for both database implementations, but I would rather let it flexible + use the badger write batch. Feel free to give it a stab in a draft PR if you want, but I couldn't make it work

File: internal/database/badger/race_test.go
Comment (kishansagathiya) [inline internal/database/badger/race_test.go:20]: I think we should be able to use just one wait group. Let me know what I might be missing.

File: internal/database/badger/settings.go
Comment (kishansagathiya) [reply internal/database/badger/settings.go:17]: why do  we need to check if it is set or unset?

**Review by kishansagathiya — CHANGES_REQUESTED** [review]
Considering we have two database implementations, we should have the same test testing for both the implementations.

File: internal/database/badger/race_test.go
Comment (qdm12) [reply internal/database/badger/race_test.go:20]: We want to wait for all goroutines (aka workers) to be ready to make sure they all operate at the same time (since starting a goroutine takes a few milliseconds) using `startWg`, and then we want to wait for all of them to exit (`endWg`)

**Comment by qdm12** [comment]
> It not nice that none of these functions are not getting used in this PR. Ideally, it will be nice if we have a db interface that is getting used. Then, In the same PR we can implement a db and plug it will rest of the codebase to use it to test if things are working fine. Let me know if that can't be done at the moment.

Yeah I get your point; using this newer interface/packages is blocked by #2831 which I'm working on now. Once this is merged, I'll rebase this branch on it and make a separate PR based on this one, which we can merge into this PR. Trying to keep this PR and the future interface-replacement PR separate to the amount of deltas in each (for reviewers), but we should merge it as a single commit in development indeed 👍

File: internal/database/badger/settings.go
Comment (qdm12) [reply internal/database/badger/settings.go:17]: To be able to set defaults if it's left unset (and it extends beyond eventually, such as merging/overriding settings structs)

File: internal/database/badger/database_test.go
Comment (qdm12) [reply internal/database/badger/database_test.go:15]: I really don't see the need for comments in this test.
Method names are quite explicit and it's tiny/meaningless values that any programmer can understand.
Comments would bloat more the test code without bringing more than what one can understand from just reading the code.

For example

```go
		err := db.Set([]byte{1}, []byte{2})
		require.NoError(t, err)
```

versus

```go
		// Set key []byte{1} with value []byte{2}
		err := db.Set([]byte{1}, []byte{2})
		require.NoError(t, err)
```

the comment is rather redundant/unneeded.

File: internal/database/badger/database.go
Comment (timwu20) [inline internal/database/badger/database.go:21]: nit:
```suggestion
func NewDatabase(settings Settings) (database *Database, err error) {
```

File: internal/database/badger/database.go
Comment (timwu20) [inline internal/database/badger/database.go:12]: nit: it's not obvious the package name in this folder is still `badger`
```suggestion
	badger "github.com/dgraph-io/badger/v3"
```

File: internal/database/badger/database.go
Comment (timwu20) [inline internal/database/badger/database.go:17]: nit: embed?
```suggestion
	*badger.DB
```

File: internal/database/badger/helpers.go
Comment (timwu20) [inline internal/database/badger/helpers.go:14]:
```suggestion
func newPrefixedKey(prefix, key []byte) (prefixedKey []byte) {
```

File: internal/database/badger/settings.go
Comment (timwu20) [reply internal/database/badger/settings.go:17]: I don't see why this is necessary.  IMO `InMemory` should default to false.

File: internal/database/badger/writebatch.go
Comment (timwu20) [inline internal/database/badger/writebatch.go:10]: what uses this?

File: internal/database/memory/database.go
Comment (timwu20) [reply internal/database/memory/database.go:15]: if tests can use the badger in memory version, do we really need this in memory implementation?

File: internal/database/badger/settings.go
Comment (kishansagathiya) [reply internal/database/badger/settings.go:17]: > To be able to set defaults if it's left unset (and it extends beyond eventually, such as merging/overriding settings structs).

I don't think I understand this completely.

>  it extends beyond eventually, such as merging/overriding settings structs

May be elaborate this part. Do we need to merge or override this setting at some point?

What I am really not able to understand is that why can't we treat false as unset. 

Say someone assigned a false value to inMemory, why do we care if it was set or it is there because it is default value?

File: internal/database/badger/database.go
Comment (qdm12) [reply internal/database/badger/database.go:21]: I think it's fine as `New` since the package is only dedicated to the badger database (and also the signature can be checked). A bit like `log.New` instead of `log.NewLogger`

File: internal/database/badger/database.go
Comment (qdm12) [reply internal/database/badger/database.go:17]: I prefer to avoid embedding if it's not needed because:

- it reduces method and field name conflicts
- it exports `DB` as a field accessible outside (big no no in that case)
- it's more cumbersome to assert against

File: internal/database/badger/settings.go
Comment (qdm12) [reply internal/database/badger/settings.go:17]: > What I am really not able to understand is that why can't we treat false as unset.

Because `false` or `true` are set values. The user can specify `false`.
Say we want to read from two configuration sources such as flags and toml, and we want flags to take precedence.
The flag specifies `--inmemory=false` and the toml config specifies `inmemory = true`. If we consider `false` as unset, then we would end up with a final value of `true` whereas it should be `false` due to precedence.
Same applies with merging/overriding settings.

It's perhaps out of scope and not used yet, but it's good code writing imo and there is no point neglecting this.

Comment (kishansagathiya) [reply internal/database/badger/settings.go:17]: > If we consider false as unset, then we would end up with a final value of true whereas it should be false due to precedence

If I want flag to take precedence, I would just take flag value and assign it to inMemory. It doesn't matter if it is set or not. So, if the flag says false, the value of inMemory will become false.

The way we deal with such setting is, 
- initialise with default settings with, say `Default()`.
- Override with say toml config (we don't need to know existing values in settings here)
- Override with say flag (we don't need existing values here as well)

I don't think we need inMemory to be pointer here.

File: internal/database/badger/helpers.go
Comment (qdm12) [reply internal/database/badger/helpers.go:14]: Done

File: internal/database/badger/database.go
Comment (qdm12) [reply internal/database/badger/database.go:12]: Done

File: internal/database/badger/writebatch.go
Comment (qdm12) [reply internal/database/badger/writebatch.go:10]: The database `NewWriteBatch()` method and the database's table method `NewWriteBatch()`

File: internal/database/memory/database.go
Comment (qdm12) [reply internal/database/memory/database.go:15]: I would like to keep it as a base for another future implementation. For example `badger` uses the badger/v3 batch implementation, but some other implementations might need to implement their own batching, and it's already there in the (zero-dependency) memory implementation. I can create an issue to remove it later on once we have at least another implementation?

**Comment by timwu20** [comment]
I think we should wait to merge this into `development`.  We should probably do some testing to ensure things are working as expected before merging in.  Can you create a release branch that will contain this PR and the PR that utilises this code?

File: internal/database/badger/settings.go
Comment (timwu20) [reply internal/database/badger/settings.go:17]: I don't see why we would want to support detecting an unset value for this attribute.  I see no reason why the default value for `InMemory` should not be `false`.

Comment (qdm12) [reply internal/database/badger/settings.go:19]: Read https://github.com/ChainSafe/gossamer/pull/3005#discussion_r1071986475 again. Empty value should be invalid (and for a boolean, false and true are both 'set values').

> If I want flag to take precedence, I would just take flag value and assign it to inMemory. It doesn't matter if it is set or not. So, if the flag says false, the value of inMemory will become false.

Say the flag is `--inmemory=false`, then your field is set to false. Then your next source (i.e. toml) sets `inmemory = true`. How does it know if it should override this `false` with its `true`?? You can have ugly logic for every field to do this, but it's much more modular to just use a boolean pointer and if it's nil, then just set it to the value read from the settings source.

File: internal/database/memory/database.go
Comment (qdm12) [reply internal/database/memory/database.go:126]: It's fully replaced in #3088 now 😉 (based on this PR)

**Pull request closed without merging**
